scripts/ci/dag.py

```diff
@@ -0,0 +1,200 @@
+                # execute it, then put it in the done queue.
+                while not shutdown.is_set():
+                    try:
+                        node = task_queue.get_nowait()
+                        state._start(node)
+                        f(node)
+                        done_queue.put(node)
+                    except Empty:
+                        time.sleep(0)  # yield to prevent busy-looping
+                        continue
+
+            for n in range(0, num_workers):  # start all workers
+                p.submit(worker, n)
+
+            tokens = max_tokens
+            num_in_progress = 0
+            last_refill = time.time()
+            while not state._is_done():
+                # This loop has two parts, `push` and `pull`.
+                #
+                # The `push` loop attempts to push tasks
+                # onto the `task_queue` while there are
+                # some tasks ready to go, and some tokens left.
+                # It is also responsible for refreshing the bucket.
+                #
+                # The `pull` loop attempts to retrieve done
+                # tasks and decrement the dependency counter on
+                # their dependents.
+                #
+                # Once a node has no pending dependencies left,
+                # it becomes ready and will be queued in one of
+                # the iterations of the `push` loop.
+                #
+                # It's important to always use non-blocking `get`
+                # with the task queue and done queue, so that both
+                # the push and pull loops can eventually make progress.
+                while len(state._queue) > 0:  # push loop
+                    now = time.time()
+                    if now - last_refill > refill_interval_sec:
+                        tokens = max_tokens - num_in_progress
+                        last_refill = now
+
+                    if len(state._queue) == 0 or tokens == 0:
+                        break
+
+                    tokens -= 1
+                    num_in_progress += 1
+                    task_queue.put(state._queue.pop())
+
+                try:
+                    while True:  # pull loop
+                        state._finish(done_queue.get_nowait())
+                        num_in_progress -= 1
+                except Empty:
+                    time.sleep(0)  # yield here to prevent busy-looping
+
+            shutdown.set()
+            state._sanity_check()
+
+
+class _NodeState(Generic[_T]):
+    def __init__(self, node: _T):
+        self.node = node
+        self.started: bool = False
+        """Whether or not a worker ever picked up this node for processing."""
+        self.counter: int = 0
+        """The number of this node's dependencies which have not yet been processed."""
+        self.dependents: list[_NodeState[_T]] = []
+        """The list of dependents which are waiting for this node to be processed."""
```

Comment:
when mixing Rust and Python like we do it is extremely confusing to read doc-comments interleaved with code

```suggestion
        self.node = node
        
        self.started: bool = False
        """Whether or not a worker ever picked up this node for processing."""
        
        self.counter: int = 0
        """The number of this node's dependencies which have not yet been processed."""
        
        self.dependents: list[_NodeState[_T]] = []
        """The list of dependents which are waiting for this node to be processed."""
```

Reply:
`counter` could be called something like `pending_dependencies`.
`counter` is about as non-descriptive as `integer` or `number`.